### PR TITLE
Support React 18

### PR DESCRIPTION
**Pull request by snowystinger**
**What**:

Start a branch to support React 18

**Why**:

We use this library to test our hooks and we're looking to support React 18

**How**:

I've changed it over according to https://github.com/reactwg/react-18/discussions/5

**Checklist**:

There are definitely broken tests :) I'm just starting to look at things, if I'm not responsive enough, feel free to take the PR over.

areas I've identified as problematic:
`dom` rendering will throw two copies of an error https://github.com/testing-library/react-hooks-testing-library/pull/655/files#diff-ce8cfe5650c6baabe7c7d381598ab9fa639f76febd230fc7e04befff5759d8e4R26
`async` I've no idea what to do about this one right now
`act` may be problematic https://github.com/reactwg/react-18/discussions/23

- [ ] Documentation updated
- [ ] Tests
- [ ] Ready to be merged
      
- [ ] Added myself to contributors table

### CODE REVIEW

**Comment by snowystinger** [comment]
On a side note, noticed this library makes it really hard to support multiple versions of React because of the lifecycle `prepare` which causes devDeps to be downloaded to people using the library's packages like a normal dependency
is there any way to get rid of that `prepare` and just create everything ahead of time?

**Comment by mpeyper** [comment]
You mean the `npm` script? I'm not fussed if you want to remove the `prepare` script as part of this if it helps.

I'm a bit short on time at the moment to look at this very deeply, but thanks for the effort you're putting in.  I'll try to take a look over the weekend.

**Comment by snowystinger** [comment]
Yeah, running against multiple versions of React would be great, we do that in our library, I'll see what I can do today

File: package.json [hunk 1]
```diff
@@ -33,7 +33,6 @@
   "scripts": {
     "setup": "npm install && npm run validate -s",
     "validate": "kcd-scripts validate",
-    "prepare": "npm run build",
```
Comment (snowystinger) [inline package.json:36]: I'm not sure what else we'll need to do now that I've removed this because of https://docs.npmjs.com/cli/v7/commands/npm-install
> If the package being installed contains a prepare script, its dependencies and devDependencies will be installed, and the prepare script will be run, before the package is packaged and installed.

File: .github/workflows/validate.yml [hunk 2]
```diff
@@ -17,6 +17,7 @@ jobs:
     strategy:
       matrix:
         node: [12.13, 12, 14, 16]
+        react: [^16.9.0, ^17.0.2, next]
```
Comment (mpeyper) [inline .github/workflows/validate.yml:20]: Can we make this `[16.9.0, ^16, ^17, next]` so we ensure it works with our minimum version, latest for each supported version and the next version?

Comment (snowystinger) [reply .github/workflows/validate.yml:20]: I'm not sure why the latest test run failed on CI

Comment (mpeyper) [reply .github/workflows/validate.yml]: It's failing coverage and then there's a thing to cancel running jobs when on job fails

```txt
[test] ------------------------|---------|----------|---------|---------|-------------------
[test] File                    | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
[test] ------------------------|---------|----------|---------|---------|-------------------
[test] All files               |   88.26 |    89.74 |   85.71 |   87.67 |                   
[test]  src                    |     100 |      100 |     100 |     100 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |     100 |      100 |     100 |     100 |                   
[test]  src/core               |     100 |      100 |     100 |     100 |                   
[test]   asyncUtils.ts         |     100 |      100 |     100 |     100 |                   
[test]   cleanup.ts            |     100 |      100 |     100 |     100 |                   
[test]   console.ts            |     100 |      100 |     100 |     100 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]  src/dom                |   63.64 |       50 |   53.85 |   63.64 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |      60 |       50 |   53.85 |      60 | 20-34             
[test]  src/helpers            |     100 |      100 |     100 |     100 |                   
[test]   createTestHarness.tsx |     100 |      100 |     100 |     100 |                   
[test]   error.ts              |     100 |      100 |     100 |     100 |                   
[test]   promises.ts           |     100 |      100 |     100 |     100 |                   
[test]  src/native             |     100 |      100 |     100 |     100 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |     100 |      100 |     100 |     100 |                   
[test]  src/server             |   57.78 |       50 |   52.94 |   57.78 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |   55.81 |       50 |   52.94 |   55.81 | 23-57             
[test]  src/types              |       0 |        0 |       0 |       0 |                   
[test]   index.ts              |       0 |        0 |       0 |       0 |                   
[test]   react.ts              |       0 |        0 |       0 |       0 |                   
[test] ------------------------|---------|----------|---------|---------|-------------------
[test] Jest: "global" coverage threshold for statements (100%) not met: 88.26%
[test] Jest: "global" coverage threshold for branches (100%) not met: 89.74%
[test] Jest: "global" coverage threshold for lines (100%) not met: 87.67%
[test] Jest: "global" coverage threshold for functions (100%) not met: 85.71%
[test] 
[test] Test Suites: 25 passed, 25 total
[test] Tests:       298 passed, 298 total
[test] Snapshots:   0 total
[test] Time:        30.323 s
[test] Ran all test suites.
[test] npm run test --silent -- --coverage exited with code 1
Error: Process completed with exit code 1.
```

Comment (mpeyper) [reply .github/workflows/validate.yml]: Hmm... This is interesting because we can't cover both sides in a single run, but we will cover all sides across all the jobs

Comment (snowystinger) [reply .github/workflows/validate.yml]: ah, I see it, hmm i'm not sure how we'd solve for that because we won't cover those lines depending on the React version we're using

Comment (mpeyper) [reply .github/workflows/validate.yml]: Drop the coverage threshold to 80% for now. I've got an idea, but will need to get some time in front of my editor to work it out.

**Comment by snowystinger** [comment]
Btw, not sure if you want to do something like this https://github.com/testing-library/react-testing-library/pull/937
TLDR; use createRoot by default but allow for opt out `render(element, { legacyRoot: true })`

File: .github/workflows/validate.yml (hunk 2 above)
Comment (snowystinger) [reply .github/workflows/validate.yml]: heh, got unlucky since the `next` tests don't pass right now that they canceled all the others, maybe those can be changed to avoid canceling the rest

Comment (mpeyper) [reply .github/workflows/validate.yml]: That would be ideal if the `next` tests are allowed to fail, but the others are strict.

**Comment by mpeyper** [comment]
I think having a `{ legacyRoot: true }` option is a good idea if you want to include it in these changes

**Comment by snowystinger** [comment]
Sure, I'll try to find some time to add it this weekend
Also, feel free to push to my branch

File: .github/workflows/validate.yml (hunk 2 above)
Comment (mpeyper) [reply .github/workflows/validate.yml]: I've done what was I was thinking of doing (`createLegacyRoot`). It reduced the amount of code that is uncovered depending on which react version is used, to the point where I'm fairly comfortable ignoring those lines as the logic is quite light around it.

I also introduced the `react/next` and `react-dom/next` types, however this broke the `node` and `jest` types ([and any other global types](https://www.typescriptlang.org/tsconfig#types) which I'm not thrilled with the fix for (also explicitly including those types).  If anyone knows a better way of adding those types them please let me know, otherwise this if fine for now.

I haven't looked into why the the async (are related) tests are failing with react 18 yet.

**Comment by mpeyper** [comment]
Re: the async tests timing out, it appears as though we are running into the same issue as [described here](https://github.com/reactwg/react-18/discussions/23#discussioncomment-812450) by @eps1lon who has also made [changes to their act functionality in the RTL PR](https://github.com/testing-library/react-testing-library/pull/937/files#diff-a4177cd5406a6c3f32827e85977bfadba791088e113656085bf736a5000ff8d2.).  I'm not sure if there was ever a resolution to that discussion as it's still very unclear to me what we are supposed to do in this situation.

If `act` is removed from the async utils, the test pass again, but the console is littered with `async act` warnings, so that is not a suitable solution.

**Comment by snowystinger** [comment]
ah :( yeah I'd seen that comment but wasn't sure how it played a part yet. Thanks for clearing that up. Unfortunately, it would seem you need to be part of their working group to upvote the discussion. Thanks for the nice little refactor as well. I didn't have time on the weekend, but I'll add `{ legacyRoot: true }` today.

**Comment by snowystinger** [comment]
Well, I didn't get to it today, this is the direction I was headed though.

createRoot.js
```
export function createRoot(container: Element) {
  let root
  let newRoot = {
    render: (hook, opts) => {
      root = (ReactDOM.createRoot && !(opts && opts.legacyRoot) ? ReactDOM.createRoot : createLegacyRoot)(container).render(hook)
      return newRoot
    },
    rerender: (hook) => {
      if (root) {
        return root.rerender(hook)
      }
    },
    unmount: () => {
      if (root) {
        return root.unmount()
      }
    },
    act: (args) => {
      if (root) {
        return root.act(args)
      }
    }
  };
  return newRoot
}

// similar for hydrate
```

dom/pure.js
```
...
  const root = createRoot(container)
  return {
    render(props?: TProps, opts?: {legacyRoot?: boolean}) {
      act(() => {
        root.render(testHarness(props), opts)
      })
    },
...
```

**Comment by mpeyper** [comment]
I think I'd want the root to returned from the `createRoot` function to match the react 18 root interface.  

As a side note, I don't think the `act` issue will have a resolution anytime soon, so would you like to put together a PR for your build changes to run against against versions of react we do currently support (`16.9.0`, `^16` and `^17`)?

**Comment by snowystinger** [comment]
Sure, sorry for the delay, we've been preparing for a release and it's been taking most of my time this week. It does appear though that there will be a resolution somewhat soon. They seem to have a direction they are going to take https://github.com/reactwg/react-18/discussions/23#discussioncomment-1087260
specifically, we'll just have to put up with the `act` warnings everywhere until they remove them :-/ but I'll break out the other stuff today and we can just leave 18 in this PR to wait for a bit longer

**Comment by mpeyper** [comment]
Oh, thanks for pointing out that discussion had advanced.

With that in mind, we will probably need some kind of `compat` layer for calling `act` in the async utils that does the wrapping in react 16/17 but not in 18 as I don't want the warning appearing for existing users that haven't upgraded to react 18 yet.

I'm not sure on the implementation of that either as the way I understand those messages, `act` will still exist and it's just the situation to use it has changed. Also, is it only when using the concurrent root, or will we need to revert to react 17 behaviour of using the legacy root in react 18?

I really wish I was part of that working group because I've got some questions about when it will warn and when it won't.

**Comment by eps1lon** [comment]
> I really wish I was part of that working group because I've got some questions about when it will warn and when it won't.

Feel free to post them via testing-library discord (and ping me). I can forward them

**Comment by brainkim** [comment]
Is this code released on npm anywhere? Just checking, no pressure.

**Comment by mpeyper** [comment]
Hi @brainkim, no, not yet. If we can get something that passes the CI tests for all variations, then I'm happy to put out an alpha build.

It's been a while since I looked at this, but I think it was all the async tests that were timing out due to a change in the `act` implementation.

**Comment by snowystinger** [comment]
hey, just catching up with all the discussions and specifically https://github.com/testing-library/react-hooks-testing-library/pull/688#issuecomment-982549094
is there anything that would be good for me to tackle here? or should we close this PR?

**Comment by mpeyper** [comment]
@snowystinger I'm happy to keep it open if anyone wants to tinker away at it, but it's looking like this library will be no more once the `@testing-library/react` and `@testing-library/react-native` have stable versions of `renderHook` available ([context](https://github.com/testing-library/react-testing-library/pull/991)) which should occur before React 18 is officially released.

**Comment by snowystinger** [comment]
Sure, happy to leave it open as well. I figured that was the outcome after reading through the other threads. Thanks for the help!

**Comment by mpeyper** [comment]
I’m closing this now to prevent confusion about what arcade doing with React 18.

Please see [the note in the README](https://github.com/testing-library/react-hooks-testing-library) for more context.

**Comment by ralphstodomingo** [comment]
I must say, I read the discussion on `react-testing-library` and I would like to ask that you reconsider deprecating this package. 

The experience I had trying to grok what differs between the `renderHook` implementations here and there, borderline horrible. I'm tearing my hair. Issues filed over there asking about the functionality missing e.g. `waitFor-` ones get only the reply that `waitFor`'s going to supposedly solve the problems, but there are no concrete documented patterns to be found. 

I'm even contemplating moving back to React 17 on a project just because of this - how is it considered acceptable these days to alienate a lot of library consumers by asking them to add much more boilerplate to do the same things that used to work before?

**Comment by mpeyper** [comment]
Hi @ralphstodomingo,

I’m so sorry it is causing you so much frustration. It’s not complete, but you can check out the [work-in-progress migration guide](https://github.com/testing-library/react-hooks-testing-library/blob/chore/migration-guide/MIGRATION_GUIDE.md) I’ve been working on for far too long which may help get you over some humps.

Ultimately, a large part of the decisions made in the other thread was to avoid this situation in the future by moving to a common api with RTL. If you’ve read that issue than you would know that there is much about that decision that I disagree with, however, I also believe that having two competing `renderHook` apis is worse for the community in the long term than having a one time upgrade pain now. I know this will be of little comfort to you or others struggling with the migration now though.
